### PR TITLE
DDPG Algorithm and Action Networks

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
+ Unit test
+ Tested on pendulum (```ddpg_pendulum.gin```), performance matches the tf version

### CODE REVIEW

File: alf/utils/dist_utils.py [hunk 1]
```diff
@@ -17,11 +17,45 @@
...
+        """
+        super(OUProcess, self).__init__()
+        self._damping = damping
+        self._stddev = stddev
+        self._x = initial_value
+        self._x.requires_grad = False
+
+    def forward(self):
+        noise = torch.zeros_like(self._x)
+        noise.normal_(0, self._stddev)
```
Comment (emailweixu) [inline alf/utils/dist_utils.py]: noise = torch.randn_like(self._x) * self._stddev

File: alf/algorithms/ddpg_algorithm_test.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        eval_env.reset()
+        for _ in range(100):
+            alg.train_iter()
+
+        eval_env.reset()
+        eval_time_step = unroll(eval_env, alg, steps_per_episode - 1)
+        print(eval_time_step.reward.mean())
+
+        self.assertAlmostEqual(
+            1.0, float(eval_time_step.reward.mean()), delta=5e-1)
```
Comment (emailweixu) [inline alf/algorithms/ddpg_algorithm_test.py]: delta of 0.5 is too loose. a constant action of 0.5 can satisfy this condition.

File: alf/algorithms/ddpg_algorithm_test.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            clear_replay_buffer=False,
+            num_envs=num_env,
+        )
+        env_class = PolicyUnittestEnv
+        steps_per_episode = 13
+        env = env_class(
+            num_env, steps_per_episode, action_type=ActionType.Continuous)
+
+        eval_env = env_class(
+            num_env, steps_per_episode, action_type=ActionType.Continuous)
```
Comment (emailweixu) [inline alf/algorithms/ddpg_algorithm_test.py]: use a large batch for eval_env can make the test more robust.

File: alf/algorithms/ddpg_algorithm.py [hunk 4]
```diff
@@ -204,29 +199,30 @@ def _critic_train_step(self, exp: Experience, state: DdpgCriticState):
...
-            q_value, critic_state = self._critic_network(
-                (exp.observation, action), network_state=state.critic)
+        q_value, critic_state = self._critic_network((exp.observation, action),
+                                                     state=state.critic)
 
-        dqda = tape.gradient(q_value, action)
+        dqda = nest.pack_sequence_as(
+            action,
+            list(
+                torch.autograd.grad(q_value, nest.flatten(action),
```
Comment (emailweixu) [inline alf/algorithms/ddpg_algorithm.py]: This is easier to read:
```python
torch.autograd.grad(q_value.sum(), nest.flatten(action))
```

File: alf/examples/ddpg_pendulum.gin [hunk 5]
```diff
@@ -27,22 +26,20 @@ DdpgAlgorithm.critic_optimizer=@critic/Adam()
...
+TrainerConfig.whole_replay_buffer_training=False
 TrainerConfig.clear_replay_buffer=False
-TrainerConfig.algorithm_ctor=@Agent
-TrainerConfig.checkpoint_interval=100000
+TrainerConfig.algorithm_ctor=@DdpgAlgorithm
 TrainerConfig.num_iterations=20000
-TrainerConfig.evaluate=1
-TrainerConfig.eval_interval=1000
-TrainerConfig.debug_summaries=0
+TrainerConfig.num_checkpoints=5
```
Comment (emailweixu) [inline alf/examples/ddpg_pendulum.gin]: Let's make a ddpg.gin to set default arguments for DDPG and people can include it to set the right config.

File: alf/utils/spec_utils.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    Returns:
+        spec_means (Tensor): the mean value of the spec bound.
+        spec_magnitudes (Tensor): the magnitude of the spec bound.
+    """
+
+    spec_means = (spec.maximum + spec.minimum) / 2.0
+    spec_magnitudes = (spec.maximum - spec.minimum) / 2.0
+    return torch.tensor(spec_means).to(
+        torch.float32), torch.tensor(spec_magnitudes).to(torch.float32)
```
Comment (hnyu) [inline alf/utils/spec_utils.py]: Should be better using dtype of the spec.

File: alf/utils/spec_utils.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+
+def scale_to_spec(tensor, spec: BoundedTensorSpec):
```
Comment (hnyu) [inline alf/utils/spec_utils.py]: I think more generally the tensor can be unbounded and we apply tanh inside this function, as also needed by the projection network https://github.com/HorizonRobotics/alf/blob/2571ee85a59b58765c8a577a28fde25381625f77/alf/networks/projection_networks.py#L144

File: alf/utils/spec_utils.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+def clip_to_spec(value, spec: BoundedTensorSpec):
+    """Clips value to a given bounded tensor spec.
+    Args:
+        value: (tensor) value to be clipped.
+        spec: (BoundedTensorSpec) spec containing min and max values for clipping.
+    Returns:
+        clipped_value: (tensor) `value` clipped to be compatible with `spec`.
+    """
+    return torch.max(
+        torch.min(value, torch.tensor(spec.maximum)),
```
Comment (hnyu) [inline alf/utils/spec_utils.py]: tensor -> as_tensor

File: alf/utils/spec_utils.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+def scale_to_spec(tensor, spec: BoundedTensorSpec):
+    """Shapes and scales a batch into the given spec bounds.
+
+    Args:
+        tensor: A [batch x n] tensor with values in the range of [-1, 1].
+        spec: (BoundedTensorSpec) to use for scaling the input tensor.
+
+    Returns:
+        A batch scaled the given spec bounds.
+    """
```
Comment (hnyu) [inline alf/utils/spec_utils.py]: It's better to first `get_outer_rank` on the tensor with the `spec` for shaping checking?

File: alf/networks/actor_networks.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+            input_tensor_spec=self._processed_input_tensor_spec,
+            conv_layer_params=conv_layer_params,
+            fc_layer_params=fc_layer_params,
+            activation=activation)
+
+        self._lstm_encoding_net = LSTMEncodingNetwork(
+            input_tensor_spec=self._encoding_net.output_spec,
+            hidden_size=lstm_hidden_size,
+            post_fc_layer_params=actor_fc_layer_params,
+            activation=activation)
```
Comment (hnyu) [inline alf/networks/actor_networks.py]: Can just combine the two encoding networks into one LSTMEncodingNetwork

File: alf/utils/dist_utils.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/utils/dist_utils.py]: Changed

File: alf/algorithms/ddpg_algorithm_test.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/algorithms/ddpg_algorithm_test.py]: Changed to a more strict test case.

File: alf/algorithms/ddpg_algorithm_test.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/algorithms/ddpg_algorithm_test.py]: The eval_env batch is increased in the test case.

File: alf/examples/ddpg_pendulum.gin (hunk 5 above)
Comment (Haichao-Zhang) [reply alf/examples/ddpg_pendulum.gin]: Created a ddpg.gin for sharing and importing by other ddpg-based applications.

File: alf/utils/spec_utils.py (hunk 6 above)
Comment (Haichao-Zhang) [reply alf/utils/spec_utils.py]: Changed.

File: alf/utils/spec_utils.py (hunk 8 above)
Comment (Haichao-Zhang) [reply alf/utils/spec_utils.py]: changed

File: alf/utils/spec_utils.py (hunk 9 above)
Comment (Haichao-Zhang) [reply alf/utils/spec_utils.py]: added

File: alf/networks/actor_networks.py (hunk 10 above)
Comment (Haichao-Zhang) [reply alf/networks/actor_networks.py]: merged into LSTMEncodingNetwork

File: alf/utils/spec_utils.py (hunk 7 above)
Comment (hnyu) [reply alf/utils/spec_utils.py]: Can you also help change the code quoted above to use this helper function?

Comment (Haichao-Zhang) [reply alf/utils/spec_utils.py]: Working on it.

Comment (Haichao-Zhang) [reply alf/utils/spec_utils.py]: This part has been changed.

File: alf/networks/projection_networks.py [hunk 11]
```diff
@@ -136,15 +138,15 @@ def __init__(self,
...
-            self._action_means = (action_high + action_low) / 2
-            self._action_magnitudes = (action_high - action_low) / 2
+            self._action_means, self._action_magnitudes = \
+                spec_utils.spec_means_and_magnitudes(action_spec)
             # Do not transform mean if scaling distribution
             if not scale_distribution:
-                self._mean_transform = (
-                    lambda inputs: self._action_means + self._action_magnitudes
-                    * inputs.tanh())
+                self._mean_transform = functools.partial(
```
Comment (emailweixu) [inline alf/networks/projection_networks.py]: frankly speaking, I feel the previous version is better because of its readability.

Comment (Haichao-Zhang) [reply alf/networks/projection_networks.py]: Changed was done upon request... I can convert it back

Comment (hnyu) [reply alf/networks/projection_networks.py]: Maybe still keep the lambda here. And the other changes are fine.

Comment (Haichao-Zhang) [reply alf/networks/projection_networks.py]: Converted back for now as the it does not help much compared to the original code.

**Pull request merged**